docs/architecture/adr-107-e2e-abci++.md

```diff
@@ -153,53 +162,65 @@ implemented in `test/e2e/tests/abci_test.go` file.
 
 ```go
 func TestABCIGrammar(t *testing.T) {
-	m := fetchABCIRequestsByNodeName(t)
 	checker := grammar.NewGrammarChecker(grammar.DefaultConfig())
 	testNode(t, func(t *testing.T, node e2e.Node) {
-		reqs := m[node.Name]
-		_, err := checker.Verify(reqs)
+		if !node.Testnet.ABCITestsEnabled {
+			return
+		}
+		reqs, err := fetchABCIRequests(t, node.Name)
 		if err != nil {
-			t.Error(err)
+			t.Error(fmt.Errorf("Collecting of ABCI requests failed: %w", err))
+		}
+		for i, r := range reqs {
+			_, err := checker.Verify(r, i == 0)
+			if err != nil {
+				t.Error(fmt.Errorf("ABCI grammar verification failed: %w", err))
+			}
 		}
 	})
 }
 ```
 
-Specifically, the test first fetches all ABCI++ requests and creates a `GrammarChecker` object. Then for each
-node in the testnet it checks if a specific set of requests respects the ABCI++ 
-grammar by calling `checker.Verify(reqs)` method. If this method returns an error, the specific execution does not respect the grammar. 
+Specifically, the test first creates a `GrammarChecker` object. Then for each node in the testnet, it collects all requests logged by this node. Remember here that `fetchABCIRequests()` returns an array of slices(`[]*abci.Request`) where the slice with index 0 corresponds to the node's `CleanStart` execution, and each additional slice corresponds to the `Recovery` execution after a specific crash. Each node must have one `CleanStart` execution and the same number of `Recovery` executions as the number of crashes that happened on this node. If collecting was successful, the test checks whether each execution respects the ABCI++ 
+grammar by calling `checker.Verify(r, i == 0)` method. The second parameter (`i == 0`) indicates whether the set of requests `r` represents a `CleanStart` or a `Recovery` execution. If `Verify` returns an error, the specific execution does not respect the grammar, and the test will fail. 
+
+The tests are executed only if `ABCITestsEnabled` is set to `true`. This is done through the manifest file. Namely, if we want to test whether CometBFT respects ABCI++ grammar, we would need to enable these tests by adding `abci_tests_enabled = true` in the manifest file of a particular testnet (e.g. `networks/ci.toml`). This will automatically activate logging on the application side. 
 
 The `Verify()` method is shown below. 
 ```go
-
-func (g *GrammarChecker) Verify(reqs []*abci.Request) (bool, error) {
-	var r []*abci.Request
-	var n int
-	if g.cfg.FilterLastHeight {
-		r, n = g.filterLastHeight(reqs)
-		if n != 0 {
-			debugMsg := fmt.Sprintf("Last height filtered, removed last %v abci calls out of %v.\n", n, len(reqs))
-			g.logger.Debug(debugMsg)
-		}
+func (g *GrammarChecker) Verify(reqs []*abci.Request, isCleanStart bool) (bool, error) {
+	r := g.filterRequests(reqs)
```

Comment:
Here, you are using `reqs` and validating its contents just after. Shouldn't you move this line after 196? (so that you first validate `reqs` and then use it)

Reply:
.. or should line 194 use `r` rather than `reqs`?